common/src/types/v0/store/node.rs

```diff
@@ -66,33 +165,242 @@ impl NodeSpec {
+                        // remove the cordon label
+                        self.drain_state =
+                            Some(DrainState::Draining(state.remove_cordon_label(label)));
+                    }
+                    DrainState::Drained(state) => {
+                        // remove the cordon label
+                        self.drain_state =
+                            Some(DrainState::Drained(state.remove_cordon_label(label)));
+                    }
+                },
+                None => {
+                    // should not be possible
+                }
+            }
+        } else if self.has_drain_label(label.clone()) {
+            // if the label is a drain label
+            match self.drain_state.clone() {
+                Some(ds) => match ds {
+                    DrainState::Cordoned(_) => {
+                        // should not be possible
+                    }
+                    DrainState::Draining(mut state) => {
+                        // remove the drain label
+                        // if none left, set to cordoned or none
+                        state = state.remove_drain_label(label);
+                        if state.drainlabels.is_empty() {
+                            if state.cordonlabels.is_empty() {
+                                self.drain_state = None;
+                            } else {
+                                self.drain_state = Some(DrainState::Cordoned(CordonedState::new(
+                                    state.cordonlabels,
+                                )));
+                            }
+                        } else {
+                            self.drain_state = Some(DrainState::Draining(state));
+                        }
+                    }
+                    DrainState::Drained(mut state) => {
+                        // remove the drain label
+                        // if none left, set to cordoned or none
+                        state = state.remove_drain_label(label);
+                        if state.drainlabels.is_empty() {
+                            if state.cordonlabels.is_empty() {
+                                self.drain_state = None;
+                            } else {
+                                self.drain_state = Some(DrainState::Cordoned(CordonedState::new(
+                                    state.cordonlabels,
+                                )));
+                            }
+                        } else {
+                            self.drain_state = Some(DrainState::Drained(state));
+                        }
+                    }
+                },
+                None => {
+                    // should not be possible
+                }
+            }
+        } else {
+            // return error
         }
     }
+
     /// Returns whether or not the node is cordoned.
     pub fn cordoned(&self) -> bool {
-        !self.cordon_labels.is_empty()
+        match self.drain_state.clone() {
```

Comment:
can use a reference here to avoid cloning

Reply:
fixed